cilium/cilium
# Support for FQDN based egress policies

Well, this didn't turn out as well as I'd hoped, but it does the basics.

Partially implements https://github.com/cilium/cilium/issues/4584

```release-note
Add toFQDNs rules that support DNS based egress policy
```

### Overview
The goal was to let us have something like:
```
    - toFQDNs:
        - matchName: cilium.io
```
and we would allow egress to the IPs seen by a pod when it queries DNS for that name. Note that items under ToFQDN have only one field now but will have at least an `except` field and some to manage TTL handling and caching.

The actual implementation is a little incomplete compared to the above.
It does:
- Parse the rules correctly
- Poll for the set of DNS names in a controller. The lookups come from the agent and this means that it can't resolve in-cluster names (when in host-mode, as a pod it will).
- When a DNS lookup returns different IPs, it updates rules in the policy repository, triggering regenerations. Note: we use the standard, built-in resolver, and there may be a cache somewhere that honors TTLs.
- ToFQDN counts as an egress policy so you must also allow egress to the DNS servers (via ToEndpoints or something).

It does not:
- intercept a pod's DNS traffic
- honour any TTL
- ensure the IPs seen by the controller are the same ones in a DNS response to a pod (this is bad!)

### Order of Operations
The general usage is:
1- Mark toFQDNs rules at the start of daemon.policyAdd.
2- Add rules to repository (including CIDR identity allocations)
3- If successful, start polling for the DNS names in all toFQDNs being added
4- For each poll where the IPs change, we will generate CIDR rules and replace the rules in the repository via a policyAdd

### Review notes
I'm not pleased with the implementation but there were a number of things that forced me down this path:
- We would prefer to not change the datapath
- We must correctly support CNAMEs (or DNS names that point to the same IPs).
- Ideally, no cross-endpoint leaks. e.g. EP 1 can access foo.com, EP 2 bar.com but they both point to the same IP
- The egress work should be extendable for ingress
- We're relying on some identity to do the policy limiting, so how we determine that identity must be stable as the DNS lookup IPs change, hence using the IPs directly. Unfortunately, this means editing the rules in the policy repository instead of updating a map somewhere.
I initially tried to have an identity per DNS name but that broke down when handling CNAMEs or 2 things pointing to a single IP. It also would require an endpoint-local ipcache table which doesn't seem worth it (the global cache meant that the IP->identity mapping would flip flop between 2 names that had the same IP).

Questions for reviewers:
- Is there a better place to plumb this?
I didn't want to re-implement the CIDR identity allocation logic in `daemon.PolicyAdd` but it isn't that complex or large. I could hook in further down during policy generation, perhaps.

- The policyRevision now increments on it's own
Since the DNS updates can happen anytime, the revision will change arbitrarily. As long as we're adding/editing rules I don't think it's possible to avoid this. If we plumb somewhere else then it might be better.

- Do we need a more parallel poller?
I wanted to avoid complexity but given enough names the poller will become too slow, and updates will seem delayed.

### Next steps
I need to add more tests and docs for this new stuff. It can be merged and I'll follow up with other, hopefully smaller, PRs.
I also need to add better error handling for cases where the policyAdd on IP changes fails.
I also need to add a CLI option to set the DNS resolver IP. This is a non-trivial addition to this PR so it might be easier to review separately. I can add it in here if that's preferred (cc @aanm)
Once this PR is done I will begin working on proxying the DNS traffic directly. This sidesteps the need for the polling mechanism, and will allow us to use the exact same IPs that are returned to the pods.

**How to test (optional)**:
Actual tests are needed. If you want to try it out:
`kubectl create -f https://raw.githubusercontent.com/cilium/cilium/HEAD/examples/minikube/http-sw-app.yaml`
```
kubectl apply -f <(cat <<EOF
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "to-fqdn"
spec:
  endpointSelector:
    matchLabels:
      class: xwing
  egress:
    - toEndpoints:
      - matchLabels:
          "k8s:io.cilium.k8s.policy.serviceaccount": kube-dns
          "k8s:io.kubernetes.pod.namespace": kube-system
          "k8s:k8s-app": kube-dns
    - toFQDN:
        - matchName: cilium.io
EOF
)
```
`kubectl exec xwing -- curl -sv cilium.io` should work
`kubectl exec xwing -- curl -sv yahoo.com` should time out


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4750)
